File: MPP-Java/src/main/java/services/TourismAgencyServices.java
package services;


import domain.Excursion;
import domain.Reservation;
import domain.User;
import domain.validators.ValidationException;
import repository.ExcursionRepository;
import repository.RepositoryException;
import repository.ReservationRepository;
import repository.UserRepository;

import java.time.LocalTime;
import java.util.List;

public class TourismAgencyServices implements Services {
    private final UserRepository userRepository;
    private final ExcursionRepository excursionRepository;
    private final ReservationRepository reservationRepository;

    public TourismAgencyServices(UserRepository userRepository, ExcursionRepository excursionRepository, ReservationRepository reservationRepository) {
        this.userRepository = userRepository;
        this.excursionRepository = excursionRepository;
        this.reservationRepository = reservationRepository;
    }

   public User login(String username, String password) throws ServicesException {
        try{
            User user = userRepository.login(username, password);
            return user;
        }
        catch (RepositoryException ex){
            throw new ServicesException(ex.getMessage());
        }
   }

    @Override
    public List<Excursion> findAllExcursions() {
        return excursionRepository.findAll();
    }

    @Override
    public List<Excursion> findFilteredExcursions(String objective, LocalTime startTime, LocalTime endTime) {
        return excursionRepository.filterByObjective(objective, startTime, endTime);
    }

    @Override
    public void addReservation(String name, Long phone, Long tickets, Excursion excursion) throws ServicesException {
        Long id = excursion.getId();
        Excursion repoExcursion = excursionRepository.findOne(id);
        if(repoExcursion == null){
            throw new ServicesException("Excursion not found!");
        }
        if(repoExcursion.getSeats() < tickets){
            throw new ServicesException("Not enough available seats!");
        }
        Reservation reservation = new Reservation(name, phone, tickets, id);
        try {
            reservationRepository.save(reservation);
            repoExcursion.setSeats(repoExcursion.getSeats() - tickets);
            excursionRepository.update(id, repoExcursion);
        }
        catch (ValidationException ex){
            throw new ServicesException(ex.getMessage());
        }
    }


}

File: CSharp/CSharp/services/TourismAgencyServices.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CSharp.domain;
using CSharp.domain.validators;
using CSharp.repository;
using CSharp.Repository;

namespace CSharp.services
{
    public class TourismAgencyServices : IServices
    {
        private readonly IUserRepository userRepository;
        private readonly IExcursionRepository excursionRepository;
        private readonly IReservationRepository reservationRepository;

        public TourismAgencyServices(IUserRepository userRepository, IExcursionRepository excursionRepository, IReservationRepository reservationRepository)
        {
            this.userRepository = userRepository;
            this.excursionRepository = excursionRepository;
            this.reservationRepository = reservationRepository;
        }

        public void login(string username, string password)
        {
            try
            {
                userRepository.Login(username, password);
            }
            catch (RepositoryException ex)
            {
                throw new ServicesException(ex.Message);
            }
        }

        public List<Excursion> findAllExcursions()
        {
            return excursionRepository.FindAll().ToList();
        }

        public List<Excursion> findAllTripsFiltered(string objective, TimeSpan from, TimeSpan to)
        {
            return excursionRepository.FilterByObjective(objective, from, to).ToList();
        }

        public void addReservation(String name, long phone, long tickets, Excursion excursion)
        {
            long id = excursion.Id;
            Excursion repoExcursion = excursionRepository.FindOne(id);
            if (repoExcursion == null)
            {
                throw new ServicesException("Excursion not found!");
            }
            if (repoExcursion.Seats < tickets)
            {
                throw new ServicesException("Not enough available seats!");
            }
            Reservation reservation = new Reservation(name, phone, tickets, id);
            try
            {
                reservationRepository.Save(reservation);
                repoExcursion.Seats = repoExcursion.Seats - tickets;
                excursionRepository.Update(id, repoExcursion);
            }
            catch (RepositoryException ex)
            {
                throw new ServicesException(ex.Message);
            }
        }
    }
}